src/coreclr/utilcode/util.cpp

```diff
@@ -773,28 +753,33 @@ DWORD LCM(DWORD u, DWORD v)
     CONTRACTL_END;
 
 #if !defined(FEATURE_REDHAWK) && (defined(TARGET_AMD64) || defined(TARGET_ARM64))
-    BOOL enableGCCPUGroups = Configuration::GetKnobBooleanValue(W("System.GC.CpuGroup"), CLRConfig::EXTERNAL_GCCpuGroup);
+    USHORT groupCount = 0;
+
+    // Starting with Windows 11 and Windows Server 2022, a process is no longer restricted to a single processor group
+    // by default. Use all processor groups if the process is not affinitized to a single one.
+    if (GetProcessGroupAffinity(GetCurrentProcess(), &groupCount, NULL) || GetLastError() != ERROR_INSUFFICIENT_BUFFER)
```

Comment:
would `GetProcessGroupAffinity` on win11/svr 2022 always (fail with `ERROR_INSUFFICIENT_BUFFER` and) return 1 for groupCount?

Reply:
On Windows 11, if the process is not affinitized, `GetProcessGroupAffinity` returns the total number of processor groups available on the machine.  Windows 10 affinitizes processes to a single processor group by default, so `GetProcessGroupAffinity` returns 1 there.

Reply:
you mean it does not fail even though you passed 0 as groupCount? I thought the intention of passing in 0 is just to get the actual group count.

example - if you have 100 cores and separated into 2 cpu groups on win10, presumably this would return `ERROR_INSUFFICIENT_BUFFER` and groupCount would be 1 if the user didn't do anything to affinitize to more than 1 group? and if the user did affintize it to all cores, groupCount would be 2. what's the behavior on win11 (since it has this primary group concept, I wasn't sure if it still has the legacy cpu gruop concept)? 

on a related note, can you still make cpu groups on win11?

Reply:
> you mean it does not fail even though you passed 0 as groupCount? I thought the intention of passing in 0 is just to get the actual group count.

The intention is to obtain the number of processor groups available to the process.  When `0` is passed in the second argument, the call fails, sets the last error to `ERROR_INSUFFICIENT_BUFFER`, and sets the second argument to the number of processor groups available to the process.

> if you have 100 cores and separated into 2 cpu groups on win10, presumably this would return ERROR_INSUFFICIENT_BUFFER and groupCount would be 1 if the user didn't do anything to affinitize to more than 1 group?

Yes.

> and if the user did affintize it to all cores, groupCount would be 2.

My understanding is that the process itself must set thread affinities in order to use more than one processor group on Windows 10.  If the process already has threads in both processor groups (not a normal scenario), reported `groupCount` will be 2.

> what's the behavior on win11 (since it has this primary group concept, I wasn't sure if it still has the legacy cpu gruop concept)?

On Windows 11 all active processor groups in the system are reported back by `GetProcessGroupAffinity` unless you have affinitized the process.

> on a related note, can you still make cpu groups on win11?

Processor groups are created on Windows 11 as well.  If you use `start /affinity`, that will affinitize the process to a single processor group. (Maybe I misunderstood your question.)

Reply:
> `GetProcessGroupAffinity` returns 1 there.

I meant the call fails and returns `1` as the `groupCount`.  Sorry for the confusion.

Reply:
> On Windows 11 all active processor groups in the system are reported back by GetProcessGroupAffinity unless you have affinitized the process.

sorry, I meant to ask what the actual behavior is. in my example, I have 100 cores, prior to win11, there's no way there is only 1 group so for sure I'd get back 2+ as groupCount. what do I get back on win11 for different situations for this groupCount? I see the following scenarios in win11 -

+ user doesn't do anything to create processor groups, would I get back 1 as groupCount or 2? in other words, is the OS still creating 2 groups underneath and returning 2 for groupCount, or is it using the new "primary group" concept and returning 1 for groupCount?

+ multiple groups are created on the machine, let's say they create 3 groups. user doesn't any kind of affitinization themselves for their process, would the OS return3 for groupCount and the process has access to all 100 cores? I presume this is the case. 

+ multiple groups are created on the machine, let's say they create 3 groups. user affinitizes their process to only 1 group, in this case it should return 1 for groupCount and the process has only access to cores in that group.

Reply:
> I have 100 cores, prior to win11, there's no way there is only 1 group so for sure I'd get back 2+ as groupCount.

`GetProcessGroupAffinity` returns only the processor groups available to the process.  On Windows 7–10 it returns a single processor group (i.e. `groupCount == 1`) unless the process explicitly set thread affinities to use other processor groups.

> user doesn't do anything to create processor groups, would I get back 1 as groupCount or 2? in other words, is the OS still creating 2 groups underneath and returning 2 for groupCount, or is it using the new "primary group" concept and returning 1 for groupCount?

Windows 11 still creates the processor groups the same way as before.  Each process is assigned one of those processor groups as the "primary group" at launch.  So in this case `groupCount` will be `2`.

> multiple groups are created on the machine, let's say they create 3 groups. user doesn't any kind of affitinization themselves for their process, would the OS return3 for groupCount and the process has access to all 100 cores? I presume this is the case.

Yes.

> multiple groups are created on the machine, let's say they create 3 groups. user affinitizes their process to only 1 group, in this case it should return 1 for groupCount and the process has only access to cores in that group.

Yes.

Reply:
cool, all clear now (yep I meant if the process had access to all 100 cores on priori win11). thanks a lot for the info!